File: src/main/java/org/example/datnbbook/service/NhaXuatBanService.java
package org.example.datnbbook.service;

import org.example.datnbbook.dto.NhaXuatBanDTO;
import org.example.datnbbook.model.NhaXuatBan;
import org.example.datnbbook.repository.NhaXuatBanRepository;
import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class NhaXuatBanService {
    @Autowired
    private NhaXuatBanRepository nhaXuatBanRepository;


    public List<NhaXuatBan> getAll() {
        return nhaXuatBanRepository.findByDeletedFalse();
    }

    public NhaXuatBan getById(Integer id) {
        return nhaXuatBanRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Nhà sản xuất không tồn tại với id: " + id));
    }

    public NhaXuatBan create(NhaXuatBan sp) {
        if (sp.getMaNhaXuatBan() == null || sp.getMaNhaXuatBan().isEmpty()) {
            String nextMaNhaXuatBan = nhaXuatBanRepository.getNextSequenceValue();
            sp.setMaNhaXuatBan(nextMaNhaXuatBan);
        }
        sp.setDeleted(false); // Đảm bảo Nhà sản xuất mới không bị đánh dấu xóa
        return nhaXuatBanRepository.save(sp);
    }

    public NhaXuatBanDTO update(Integer id, NhaXuatBanDTO nhaXuatBanDTO) {
        NhaXuatBan nhaXuatBan = nhaXuatBanRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Nhà sản xuất không tồn tại với id: " + id));

        BeanUtils.copyProperties(nhaXuatBanDTO, nhaXuatBan ,"maNhaXuatBan","id", "deleted");
        nhaXuatBan = nhaXuatBanRepository.save(nhaXuatBan);
        return convertToDTO(nhaXuatBan);
    }

    public void delete(Integer id) {
        NhaXuatBan nhaXuatBan = nhaXuatBanRepository.findById(id)
                .filter(sp -> !sp.getDeleted())
                .orElseThrow(() -> new RuntimeException("Nhà sản xuất không tồn tại với id: " + id));
        nhaXuatBan.setDeleted(true);
        nhaXuatBanRepository.save(nhaXuatBan);
    }

    public List<NhaXuatBan> search(String keyword) {
        return nhaXuatBanRepository.search(keyword);
    }

    private NhaXuatBanDTO convertToDTO(NhaXuatBan nhaXuatBan) {
        NhaXuatBanDTO nhaXuatBanDTO = new NhaXuatBanDTO();
        BeanUtils.copyProperties(nhaXuatBan, nhaXuatBanDTO);
        return nhaXuatBanDTO;
    }
}

File: src/frontend/vue-notus-main/vue-notus-main/vue-notus-main/src/service/NhaXuatBanService.js
import api from './api';
import NhaXuatBan from '../models/NhaXuatBan';

class NhaXuatBanService {
  async getAll() {
    const response = await api.get('/nha-xuat-ban');
    return response.data.map(item => new NhaXuatBan(item));
  }

  async getById(id) {
    const response = await api.get(`/nha-xuat-ban/${id}`);
    return new NhaXuatBan(response.data);
  }

  async create(nhaXuatBan) {
    const response = await api.post('/nha-xuat-ban', nhaXuatBan);
    return new NhaXuatBan(response.data);
  }

  async update(id, nhaXuatBan) {
    const response = await api.put(`/nha-xuat-ban/${id}`, nhaXuatBan.toJSON());
    return new NhaXuatBan(response.data);
  }

  async delete(id) {
    await api.delete(`/nha-xuat-ban/${id}`);
  }

  async search(keyword) {
    const response = await api.get('/nha-xuat-ban/search', {
      params: { keyword }
    });
    return response.data.map(item => new NhaXuatBan(item));
  }
}

export default new NhaXuatBanService();